facebookresearch/multimodal
# [CI] Update build workflow for Beta release

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #334
* #326
* __->__ #322
* #323
* #333



Summary:

Refactor existing steps:
- Set `bash` as the default shell to run all steps; Customized with additional options
  - Remove the `-u` option that "treats unset variables as an error when substituting" as this can disrupt activating the correct conda environment
  - In general, `-u` can create unwanted noise due to system reasons unrelated to the job, see this [thread](https://bbs.archlinux.org/viewtopic.php?id=241893).
- Use conda action to set up miniconda, activate the conda environment; Add sanity check conda/python
- Add steps to update `pip`, `wheel`, `conda`
- Use `python -m $exec` to ensure the right version of `$exec` is invoked for the current environment (except for `pytest`)
  - `python -m pytest` automatically adds the current working directory to `sys.path`. This behavior is undesirable
    for testing the installed package rather than the source code.
- Add `package_name`, `python_tag` to wheel info; Add argument parsing in `setup.py` to process package info
- Update `requirements.txt`: remove dependency on `packaging`, add `DALL-E` which the core lib now takes a dep on
- Sort `dev-requirements.txt` alphabetically
- Add code cov, profiling options to `pytest`

Add the step to upload the binary to PyPI:
- Create a PyPI API token and store it amongst the repo's credentials
- Add the step in `test_buil.yml` to publish the wheel
- Update `README.md` with new installation instructions

Test Plan:

- Verify the manually dispatched workflow is successful: look for the lastest workflow run [here](https://github.com/facebookresearch/multimodal/actions/workflows/test_build.yml)
- Verify the existence of PyPI project: https://pypi.org/project/torchmultimodal-nightly/
- `python -m pip install torchmultimodal-nightly`

Differential Revision: [D39904704](https://our.internmc.facebook.com/intern/diff/D39904704)
